### DISCUSSION TITLE
How to Connect Python Code to an Opened Mechanical Project or Launch a Specific One?

### DISCUSSION THREAD

**Discussion by thanborb** [ansys/pymechanical#996] [Q&A] [answered]
I have a static structural project in Workbench, and when I open the model, I can use Mechanical scripting to perform some calculations. However, I want to execute my script externally from VS Code instead of writing it directly in Mechanical scripting.

To achieve this, I’m first trying to connect from my Python code to my already opened Mechanical instance. I’ve attempted using the methods connect_to_mechanical and launch_mechanical, but I’m unsure if I use them correctly. I suspect I might be using the wrong port or missing a step to find the correct one.

Once the connection is established, I believe I can use the run_python_script_from_file method to execute my script externally.

My question is:

1) How can I connect Python to an already opened instance of Mechanical?
2) If that’s not possible, how can I open a specific Mechanical project from Python?
Any guidance or examples would be greatly appreciated!

**Comment by ekostson** [comment]
Hi

Can not manage also on the 1st, but perhaps @koubaa can help here.

For 2nd question:

 see here in section Launch Mechanical and import Mechanical scripting variables:

https://developer.ansys.com/blog/create-report-saved-mechanical-session-mechdb-or-mechdat

so use:

e.g., : app.open(os.path.join(work_dir,'fileacp.mechdat'))

to open an existing mechanical data file

**Comment by dipinknair** [comment, marked as answer]
Hi @thanborb . 

I am answering all your questions assuming you are using pymechanical remote session method.

For your **first** question,
If you want to connect to A mechanical instance that is already running, you need to start Grpc Server on that instance. These are the steps to start grpc server and connect.

- Execute this on already running Mechanical ``Ansys.ACT.Mechanical.MechanicalAPI.Instance.ApplicationAPI.StartGrpcServer(10000)`` where 10000 is the port number
- Connect to Mechanical using pymechanical with ``connect_to_mechanical`` using same 10000 port or alternately like [this](https://mechanical.docs.pyansys.com/version/stable/getting_started/running_mechanical.html#connect-to-a-mechanical-session).

For the **second** question,
If you have a Mechanical Project file saved as `.mechdb` or `.mechdat` then you can start new pymechanical session and run mechanical script where you can open the Mechanical project. This is possible if you integrate opening Mechanical project files in your Mechanical script which you already have by adding line `DataModel.Project.Open(<path to file>)` at start. 

Note: You can also launch Mechanical instance using `ansys-mechanical` cli which is included with Pymechanical. Ref https://mechanical.docs.pyansys.com/version/stable/getting_started/running_mechanical.html#launch-mechanical-from-the-command-line


In above comment, @ekostson provided another method that involves embedding an instance of PyMechanical, which creates a Python object of Mechanical. You can find examples of both methods below.

Remote method examples : https://examples.mechanical.docs.pyansys.com/examples/index.html
Embedding instance examples : https://embedding.examples.mechanical.docs.pyansys.com/examples/index.html

**Reply by thanborb** [reply]
Thank you very much! I successfully connected to my Mechanical instance. I have one last question: If I use the ".run_python_script_from_file(script_to_run)" method with my Mechanical instance, will the script run the same way as it does when executed within the Mechanical Scripting console (image below)? Up until now, I have been running my scripts directly in the Mechanical Scripting console, but now I want to execute them externally.
![image](https://github.com/user-attachments/assets/d33d9112-cab6-4c30-9103-c0c5e3da458c)

**Reply by dipinknair** [reply]
Yes. When you use`` run_script_from_file()``, the script run the same way as it does when executed within the Mechanical Scripting console.

**Reply by datboi179** [reply]
Hi @dipinknair ,
It's been a while since you posted this but I would like you to know you have genuinely saved my final year project :)

Do you know if there is a similar command I can use to **_close_** the server from the Mechanical terminal?

**Reply by dipinknair** [reply]
Yes, `Ansys.ACT.Mechanical.MechanicalAPI.Instance.ApplicationAPI.StopGrpcServer()`
For exiting Mechanical `ExtAPI.Applicaiton.Exit()`
